### ISSUE TITLE
Remove magic in tests

### ISSUE THREAD

**Issue by datapythonista** [ibis-project/ibis#2478]
As identified in #2477, seems like some of the tests are not being executed. One of the files is `ibis/tests/sql/test_compiler.py`.

Checking at that file, we've got the impala backend imported at the top (see [the code](https://github.com/ibis-project/ibis/blob/master/ibis/tests/sql/test_compiler.py#L9)):
```python
from ibis import impala  # noqa: E402
```

And then, we've got some calls to its `compile` method, for example this call (see [the code](https://github.com/ibis-project/ibis/blob/master/ibis/tests/sql/test_compiler.py#L202)):
```python
query = impala.compile(expr2)
```

Not sure when this stopped working, or what was changed, but this `compile` method doesn't seem to exist:
```python
(ibis-dev) [mgarcia@xps ibis]$ python
Python 3.8.6 | packaged by conda-forge | (default, Oct  7 2020, 19:08:05) 
[GCC 7.5.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from ibis import impala  # noqa: E402
>>> expr2 = None
>>> query = impala.compile(expr2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'ibis.impala' has no attribute 'compile'
```

When executing with pytest, the tests do work, and apparently the `impala` object is the `ibis/impala/api.py` module, and not the `ibis/impala/__init__.py` one, as defined in the import.

Not sure what sort of magic we have in the tests, but would be very nice to get rid of it.

**Comment by matthewmturner** [comment]
I think its because in `ibis/__init__.py` we have: 
```
with suppress(ImportError):
    import ibis.impala.api as impala
```
If you remove the `suppress(ImportError)` there is an `ImportError` when importing ibis
```
>>> import ibis
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/matth/development/ibis/ibis/__init__.py", line 32, in <module>
    import ibis.impala.api as impala  # noqa: F401
  File "/home/matth/development/ibis/ibis/impala/api.py", line 5, in <module>
    from ibis.impala.client import (  # noqa: F401
  File "/home/matth/development/ibis/ibis/impala/client.py", line 26, in <module>
    from ibis.impala.compat import HS2Error, ImpylaError, impyla
  File "/home/matth/development/ibis/ibis/impala/compat.py", line 15, in <module>
    import impala.dbapi as impyla  # noqa
ModuleNotFoundError: No module named 'impala'
```
I think its backstopping to `ibis/impala` instead of `ibis/impala/api.py` hence the error.

I know we had issues with impyla in #2435 but im still not exactly clear whats going on here.  

naively it looks were missing impyla from the environment but the import error has been suppressed so it hasnt been picked up.  im confused how these tests could have ever worked if `impala.compile` has always been used with impyla (and impyla was not part of the environment).

i guess the idea is to move to the `backends/base_sql` as the compiler?

**Comment by datapythonista** [comment]
Thanks for the research @matthewmturner. I think those are very good points, and I think they are related, but I don't think they explain why the `impala` in `from ibis import impala` ends up being the `impala/api.py` file. At least I still don't understand.

When `impyla` is installed, the example in the issue still fails, and the tests work. So, even when that `suppress` is not triggered, the error still happens.

I didn't check in detail, but I do think the `impala.compile` is reused across backends, and needs to be moved to `base_sql`. But probably better to refactor one thing at a time, I think it can become to complex to start moving things in parallel, given all the weird errors we're getting.

**Comment by matthewmturner** [comment]
There is the following which when removed results in test failure instead of skipping.

https://github.com/ibis-project/ibis/blob/fe04a29325f1316a4e6992f8369636a52c7af325/ibis/tests/sql/test_compiler.py#L14

Is the objective to remove this so that error / failure is raised and we get full test report for `tests/sql`?

**Comment by datapythonista** [comment]
> Is the objective to remove this so that error / failure is raised and we get full test report for `tests/sql`?

That makes sense, but it may be a bit more complicated. So far, what has been done is that some modules can be imported even if impyla is not installed, but the specific functions that use it fail if called. This is implemented by simply having the imports of impyla inside the functions that use it.

Those `importorskip` are related, and skip those when impyla is not installed. If the backends are independent, no backend depends on impala, there is a base backend with the shared functionality, and the architecture in general is appropiate, I don't think any of that is required. The whole `ibis/backends/impala` will only be loaded when the impala backend is used, and if impyla is not installed is fair to fail immediately.

Similar for tests. If the impala tests are in `ibis/backends/impala/tests`, they will be called when impala is tested, if impyla is not installed we should fail, and the rest of the modules shouldn't require impyla. And those `importorskip` should all go, they just cause that we stop testing things without realizing.

Having said that, and without having checked the code, I think that `importorskip` should be removed. But that will probably require that some of the tests in that file are moved to `ibis/backends/impala/tests` or somewhere else, depending on what they are testing.

It would be really great to see that fixed, I think all the nonsense errors we're getting will stop happening when we stop skipping tests and silently ignoring exceptions.

**Comment by matthewmturner** [comment]
Ok, makes sense.

For this issue should we limit it to removing the `importorskip` to at least get the `sql/tests` running and then work on a separate one for moving the tests to `ibis/backends/impala/tests`?

**Comment by datapythonista** [comment]
> For this issue should we limit it to removing the `importorskip` to at least get the `sql/tests` running and then work on a separate one for moving the tests to `ibis/backends/impala/tests`?

Not sure if removing the import alone will fix the original error of this issue. But seems like a good first step. If you want to open a PR for that, and reference this issue but not close it, that would be great.
